Drag snapshot does not paint for normal flow content

nullhook opened the pull request:
#### 1d76f585b407b4193756ebe2d18d83b230fa962e
<pre>
Drag snapshot does not paint for normal flow content
<a href="https://rdar.apple.com/154202408">rdar://154202408</a>

Reviewed by Simon Fraser.

Drag snapshots did not work for normal flow content because
enclosingLayer found the nearest layer as the HTML root layer.
As a result, the layer used for painting decisions was a much
higher ancestor. We now restrict this with a hasLayer check.

For non-layer roots, we exclude positioned descendants from the
drag image if they extend outside the containing block.

All the special inclusion/exclusion logic is now guraded behind a
new paint behavior DraggableSnapshot.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::snapshotNode):

snapshot tests only draggable paint behavior.

Canonical link: <a href="https://commits.webkit.org/297110@main">https://commits.webkit.org/297110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c040f0fe793ebac8e096140aa24f96959eb0d13d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30179 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20612 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116546 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60787 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38767 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84045 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113468 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24627 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99513 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64486 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23989 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17652 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60341 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94008 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119336 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37560 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27884 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93015 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37934 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95784 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92838 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23665 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37834 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15575 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33533 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37455 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42926 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37117 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38826 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3749,21 +3749,28 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
                 }
             } else
                 shouldPaintContent = false;
+        } else {
+            if (renderer().isAbsolutelyPositioned() && &renderer() != paintingInfo.subtreePaintRoot) {
```
```suggestion
        } else if (renderer().isAbsolutelyPositioned() && &renderer() != paintingInfo.subtreePaintRoot) {
```

lilyspiniolas commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2118,7 +2118,10 @@ ExceptionOr<RefPtr<ImageData>> Internals::snapshotNode(Node& node)
     if (!document || !document->frame())
         return Exception { ExceptionCode::InvalidAccessError };
 
+    // Replicate what createDragImageForNode does
+    // We're always testing for drag snapshots here
```
nit: add punctuation here, maybe elaborate on these comments, too.

smfr commented on Source/WebCore/platform/DragImage.cpp:
```diff
@@ -120,7 +120,11 @@ static DragImageRef createDragImageFromSnapshot(RefPtr<ImageBuffer> snapshot, No
     ScopedNodeDragEnabler enableDrag(frame, node);
-    return createDragImageFromSnapshot(snapshotNode(frame, node, { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() }), &node);
+
+    SnapshotOptions options { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() };
```
Consider:
```
 auto options =  SnapshotOptions options { { SnapshotFlags::DraggableElement }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() }; { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() };
```

nullhook replied:
there's a typo here but i get what you mean, fixed.

smfr commented on Source/WebCore/platform/DragImage.cpp:
```diff
@@ -218,7 +222,10 @@ DragImageRef createDragImageForImage(LocalFrame& frame, Node& node, IntRect& ima
     imageRect = paintingRect;
 
-    return createDragImageFromSnapshot(snapshotNode(frame, node, { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() }), &node);
+    SnapshotOptions options { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() };
```
Ditto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3749,21 +3749,28 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
 
             if (isLayerInSubtree) {
-                if (paintingInfo.subtreePaintRoot != &renderer()) {
+                if (&renderer() != paintingInfo.subtreePaintRoot) {
```
No need to flip this.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3749,21 +3749,28 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
                 shouldPaintContent = false;
+        } else {
+            if (renderer().isAbsolutelyPositioned() && &renderer() != paintingInfo.subtreePaintRoot) {
+                if (CheckedPtr rootAsBlock = dynamicDowncast<RenderBlock>(paintingInfo.subtreePaintRoot)) {
```
I think this branch also behaves correctly for the case where `paintingInfo.subtreePaintRoot->hasLayer()` is true, so consider removing the first branch.

If you do keep both, factor the `isContainingBlockAncestorFor` chunk into a lambda to avoid repetition.

smfr commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2119,6 +2119,7 @@ ExceptionOr<RefPtr<ImageData>> Internals::snapshotNode(Node& node)
         return Exception { ExceptionCode::InvalidAccessError };
 
     SnapshotOptions options { { }, ImageBufferPixelFormat::BGRA8, DestinationColorSpace::SRGB() };
+    options.flags.add(SnapshotFlags::DraggableElement);
```
Ditto.

smfr commented on LayoutTests/fast/snapshot/positioned-static.html:
```diff
@@ -0,0 +1,55 @@
+}
+
+window.addEventListener('load', () => {
+    setTimeout(main, 200);
```
Don't use a 200ms timeout. If every one of our thousands of layout tests took 200ms, they would take hours to run.

smfr commented on LayoutTests/fast/snapshot/positioned-static.html:
```diff
@@ -0,0 +1,55 @@
+        return;
+    }
+
+    const imageData = internals.snapshotNode(box);
```
Consider moving this shared JS into a `resources/snapshot-helper.js` file.

nullhook replied:
done

smfr reviewed (commented):
Please wrap the commit message to something sensible like 80 or 90 columns.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f220997b694c86be755bc1a1a7514f4c44c561ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47192

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2dc999d3e66342aef6f0ab0316da38ca5e4e37dd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47192

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3752,21 +3752,21 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
-            } else
+            if (isLayerInSubtree && (paintingInfo.subtreePaintRoot != &renderer() && shouldExcludeBasedOnContainingBlock()))
                 shouldPaintContent = false;
+        } else if (renderer().isAbsolutelyPositioned() && paintingInfo.subtreePaintRoot != &renderer() && shouldExcludeBasedOnContainingBlock()) {
```
Does it work if you get rid of the `isAbsolutelyPositioned()` here, and remove the first `paintingInfo.subtreePaintRoot->hasLayer()` branch? I don't think you should need both branches.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c040f0fe793ebac8e096140aa24f96959eb0d13d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47192

nullhook commented on LayoutTests/resources/snapshot-helper.js:
```diff
@@ -0,0 +1,26 @@
+window.SnapshotHelper = class SnapshotHelper {
+    static async takeSnapshot(node)
+    {
+        // Wait for two animation frames to ensure the node has been painted.
```
adding `updateLayoutIgnorePendingStylesheets` helped

nullhook committed:
Drag snapshot does not paint for normal flow content
rdar://154202408

Reviewed by Simon Fraser.

Drag snapshots did not work for normal flow content because
enclosingLayer found the nearest layer as the HTML root layer.
As a result, the layer used for painting decisions was a much
higher ancestor. We now restrict this with a hasLayer check.

For non-layer roots, we exclude positioned descendants from the
drag image if they extend outside the containing block.

All the special inclusion/exclusion logic is now guraded behind a
new paint behavior DraggableSnapshot.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::snapshotNode):

snapshot tests only draggable paint behavior.

Canonical link: https://commits.webkit.org/297110@main
